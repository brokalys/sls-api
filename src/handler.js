import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
