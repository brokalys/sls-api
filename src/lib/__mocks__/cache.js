export default {
  run: jest.fn((key, params, callback) => callback(params)),
  get: jest.fn(),
  set: jest.fn(),
};
