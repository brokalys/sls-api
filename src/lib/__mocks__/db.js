export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
