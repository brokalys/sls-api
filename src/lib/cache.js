import crypto from 'crypto';

import mysql from './db';

function hash(key) {
  return crypto.createHash('md5').update(JSON.stringify(key)).digest('hex');
}

export default {
  async run(key, params, callback) {
    const cached = await this.get(key, params);

    if (cached) {
      return cached;
    }

    const newData = await callback(params);
    await this.set(key, params, newData);

    return newData;
  },

  async get(key, params) {
    const response = await mysql.query({
      sql: `SELECT value FROM ${process.env.DB_CACHE_DATABASE}.cache WHERE \`key\` = ? and \`params_hash\` = ?`,
      values: [key, hash(params)],
    });

    if (response && response.length) {
      return JSON.parse(response[0].value);
    }
  },

  set(key, params, value) {
    return mysql.query({
      sql: `INSERT INTO ${process.env.DB_CACHE_DATABASE}.cache SET ?`,
      values: {
        key,
        params: JSON.stringify(params),
        params_hash: hash(params),
        value: JSON.stringify(value),
      },
    });
  },
};
