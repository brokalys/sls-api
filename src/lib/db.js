import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
