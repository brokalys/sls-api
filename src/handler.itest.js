import { createTestClient } from 'apollo-server-testing';

import { server } from 'handler';
import cache from 'lib/cache';
import db from 'lib/db';

jest.mock('lib/cache');
jest.mock('lib/db');

db.query.mockImplementation(() => [
  {
    count: 1,
    price: 110000.0,
    lat: 56.9366684,
    lng: 24.0794235,
    price_per_sqm: 100,
    area: 120,
    area_measurement: 'm2',
  },
  {
    price: 170000.0,
    lat: 56.9366684,
    lng: 24.0794235,
    price_per_sqm: 0,
    area: 1,
    area_measurement: 'ha',
  },
]);

describe('Query', () => {
  let query;

  beforeEach(() => {
    cache.get.mockReset();

    process.env.BROKALYS_PRIVATE_KEY = 'PRIVATE_KEY';

    const utils = createTestClient(server);
    query = utils.query;
  });

  describe('getPropertiesForPinger', () => {
    // @todo: passing headers to integration tests is currently not possible..
    // @see https://github.com/apollographql/apollo-server/issues/2277
    xtest('fails validation if region is invalid', async () => {
      const res = await query({
        query: `
        {
          getPropertiesForPinger(
            category: APARTMENT
            type: SELL
            start_date: "2019-01-01 00:00:00"
            region: "WRONG"
          ) {
            price
          }
        }
      `,
      });

      expect(res).toMatchSnapshot();
    });

    test('fails authenticating and returns nothing', async () => {
      const res = await query({
        query: `
        {
          getPropertiesForPinger(
            category: APARTMENT
            type: SELL
            start_date: "2019-01-01 00:00:00"
            region: "56.992294 24.136619, 56.976394 23.995790, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176, 56.992294 24.136619"
          ) {
            price
          }
        }
      `,
      });

      expect(res).toMatchSnapshot();
    });

    // @todo: passing headers to integration tests is currently not possible..
    // @see https://github.com/apollographql/apollo-server/issues/2277
    xtest('fetches data', async () => {
      const res = await query({
        query: `
        {
          getPropertiesForPinger(
            category: APARTMENT
            type: SELL
            start_date: "2019-01-01 00:00:00"
            region: "56.992294 24.136619, 56.976394 23.995790, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176, 56.992294 24.136619"
          ) {
            price
          }
        }
      `,
      });

      expect(res).toMatchSnapshot();
    });
  });
});

describe('Mutation', () => {
  let mutate;

  beforeEach(() => {
    cache.get.mockReset();

    const utils = createTestClient(server);
    mutate = utils.mutate;
  });

  describe('confirmPinger', () => {
    test('confirms an existing pinger', async () => {
      db.query.mockImplementation(() => ({
        affectedRows: 1,
      }));

      const response = await mutate({
        mutation: `
          mutation {
            confirmPinger(
              id: 1
              confirm_key: "test_123"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('fails confirming a pinger with wrong credentials, but still responds with status = true', async () => {
      db.query.mockImplementation(() => ({
        affectedRows: 0,
      }));

      const response = await mutate({
        mutation: `
          mutation {
            confirmPinger(
              id: 1
              confirm_key: "wrong_credentials"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('fails confirming with missing `id`', async () => {
      const response = await mutate({
        mutation: `
          mutation {
            confirmPinger(
              confirm_key: "test_123"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('fails confirming with missing `confirm_key`', async () => {
      const response = await mutate({
        mutation: `
          mutation {
            confirmPinger(
              id: 1
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });
  });

  describe('unsubscribePinger', () => {
    test('unsubscribes an existing pinger', async () => {
      db.query.mockImplementation(() => ({
        affectedRows: 1,
      }));

      const response = await mutate({
        mutation: `
          mutation {
            unsubscribePinger(
              id: 1
              unsubscribe_key: "test_123"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('unsubscribes all pingers of an email', async () => {
      db.query.mockReturnValueOnce([
        {
          email: 'test@brokalys.com',
        },
      ]);
      db.query.mockReturnValueOnce({
        affectedRows: 1,
      });

      const response = await mutate({
        mutation: `
          mutation {
            unsubscribePinger(
              id: 1
              unsubscribe_key: "test_123"
              all: true
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('fails unsubscribing a pinger with wrong credentials, but still responds with status = true', async () => {
      db.query.mockImplementation(() => ({
        affectedRows: 0,
      }));

      const response = await mutate({
        mutation: `
          mutation {
            unsubscribePinger(
              id: 1
              unsubscribe_key: "wrong_credentials"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('fails unsubscribing with missing `id`', async () => {
      const response = await mutate({
        mutation: `
          mutation {
            unsubscribePinger(
              unsubscribe_key: "test_123"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });

    test('fails unsubscribing with missing `unsubscribe_key`', async () => {
      const response = await mutate({
        mutation: `
          mutation {
            unsubscribePinger(
              id: 1
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });
  });
});
