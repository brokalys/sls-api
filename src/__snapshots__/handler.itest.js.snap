// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation confirmPinger confirms an existing pinger 1`] = `
Object {
  "data": Object {
    "confirmPinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation confirmPinger fails confirming a pinger with wrong credentials, but still responds with status = true 1`] = `
Object {
  "data": Object {
    "confirmPinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation confirmPinger fails confirming with missing \`confirm_key\` 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Field \\"confirmPinger\\" argument \\"confirm_key\\" of type \\"String!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation confirmPinger fails confirming with missing \`id\` 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Field \\"confirmPinger\\" argument \\"id\\" of type \\"Int!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation unsubscribePinger fails unsubscribing a pinger with wrong credentials, but still responds with status = true 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation unsubscribePinger fails unsubscribing with missing \`id\` 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Field \\"unsubscribePinger\\" argument \\"id\\" of type \\"Int!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation unsubscribePinger fails unsubscribing with missing \`unsubscribe_key\` 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Field \\"unsubscribePinger\\" argument \\"unsubscribe_key\\" of type \\"String!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation unsubscribePinger unsubscribes all pingers of an email 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation unsubscribePinger unsubscribes an existing pinger 1`] = `
Object {
  "data": Object {
    "unsubscribePinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query getPropertiesForPinger fails authenticating and returns nothing 1`] = `
Object {
  "data": Object {
    "getPropertiesForPinger": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
